pkg/api/resource/quantity.go

```diff
@@ -0,0 +1,422 @@
+	// functions are not sufficient.
+	Amount *inf.Dec
+
+	// Change Format at will. See the comment for Canonicalize for
+	// more details.
+	Format
+}
+
+// Format lists the three possible formattings of a quantity.
+type Format string
+
+const (
+	DecimalExponent = Format("DecimalExponent") // e.g., 12e6
+	BinarySI        = Format("BinarySI")        // e.g., 12Mi (12 * 2^20)
+	DecimalSI       = Format("DecimalSI")       // e.g., 12M  (12 * 10^6)
+)
+
+// Q turns the given string into a quantity or panics; for tests
+// or others cases where you know the string is valid.
+func Q(str string) *Quantity {
+	q, err := ParseQuantity(str)
+	if err != nil {
+		panic(fmt.Errorf("cannot parse '%v': %v", str, err))
+	}
+	return q
+}
+
+const (
+	// splitREString is used to separate a number from its suffix; as such,
+	// this is overly permissive, but that's OK-- it will be checked later.
+	splitREString = "^([+-]?[0123456789.]+)([eEimkKMGTP]*[-+]?[0123456789]*)$"
+)
+
+var (
+	// splitRE is used to get the various parts of a number.
+	splitRE = regexp.MustCompile(splitREString)
+
+	// Errors that could happen while parsing a string.
+	ErrFormatWrong = errors.New("quantities must match the regular expression '" + splitREString + "'")
+	ErrNumeric     = errors.New("unable to parse numeric part of quantity")
+	ErrSuffix      = errors.New("unable to parse quantity's suffix")
+
+	// Commonly needed big.Int values-- treat as read only!
+	ten      = big.NewInt(10)
+	zero     = big.NewInt(0)
+	one      = big.NewInt(1)
+	thousand = big.NewInt(1000)
+	ten24    = big.NewInt(1024)
+
+	// Commonly needed inf.Dec values-- treat as read only!
+	decZero     = inf.NewDec(0, 0)
+	decOne      = inf.NewDec(1, 0)
+	decMinusOne = inf.NewDec(-1, 0)
+	decThousand = inf.NewDec(1000, 0)
+
+	// Smallest and largest (in magnitude) numbers allowed.
+	minAllowed = inf.NewDec(1, 3)         // == 1/1000
+	maxAllowed = inf.NewDec((1<<63)-1, 0) // == max int64
+
+	// The maximum value we can represent milli-units for.
+	// Compare with the return value of Quantity.Value() to
+	// see if it's safe to use Quantity.MilliValue().
+	MaxMilliValue = ((1 << 63) - 1) / 1000
+)
+
+// ParseQuantity turns str into a Quantity, or returns an error.
+func ParseQuantity(str string) (*Quantity, error) {
+	parts := splitRE.FindStringSubmatch(strings.TrimSpace(str))
+	if len(parts) != 3 {
+		return nil, ErrFormatWrong
+	}
+
+	amount := new(inf.Dec)
+	if _, ok := amount.SetString(parts[1]); !ok {
+		return nil, ErrNumeric
+	}
+
+	base, exponent, format, ok := quantitySuffixer.interpret(suffix(parts[2]))
+	if !ok {
+		return nil, ErrSuffix
+	}
+
+	// So that no one but us has to think about suffixes, remove it.
+	if base == 10 {
+		amount.SetScale(amount.Scale() + inf.Scale(-exponent))
+	} else if base == 2 {
+		// numericSuffix = 2 ** exponent
+		numericSuffix := big.NewInt(1).Lsh(one, uint(exponent))
+		amount.UnscaledBig().Mul(amount.UnscaledBig(), numericSuffix)
```

Comment:
I think you're right-- I had done an unhealthy amount of reading of inf.Dec's code, which is full of stuff like this, which probably means I should treat it as an implementation detail.